### PR TITLE
Botnet rebase

### PR DESCRIPTION
**Pull request by sigvartmh**
Lets go haaaam

### CODE REVIEW

File: samples/nrf5340/netboot/src/main.c [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+		goto failure;
+	}
+	
+	cmd = pcd_get_cmd((void*)PCD_CMD_ADDRESS);
+	rsp = pcd_get_cmd((void*)PCD_RSP_ADDRESS);
+	if (cmd != NULL && rsp != NULL) {
+		err = pcd_transfer(cmd, rsp, fdev);
+		if (err != 0) {
+			printk("Failed to transfer image: %d. \n\r", err);
+			return;
```
Comment (sigvartmh) [inline samples/nrf5340/netboot/src/main.c]: Missing goto. Should we use `while` + `break` instead?

File: samples/nrf5340/netboot/src/main.c [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	}
+
+	err = fprotect_area(PM_APP_ADDRESS, PM_APP_SIZE);
+	if (err) {
+		printk("Failed to protect app flash: %d. \n\r", err);
+		goto failure;
+	}
+
+	bl_boot(fw_info_find(s0_addr));
+
```
Comment (sigvartmh) [inline samples/nrf5340/netboot/src/main.c:63]: maybe add early return? Since we don't really should set the fail magic

File: cmake/mcuboot.cmake [hunk 3]
```diff
@@ -181,6 +181,34 @@ if(CONFIG_BOOTLOADER_MCUBOOT)
     "version_MCUBOOT=${CONFIG_MCUBOOT_IMAGE_VERSION}"
     )
 
+  if (CONFIG_BT_RPMSG_NRF53 AND CONFIG_SOC_NRF5340_CPUAPP)
+    # The bootloader on the network core is enabled. The validation of this
```
Comment (hakonfam) [inline cmake/mcuboot.cmake:185]: Is it? IT is the inclusion of mcuboot on the app core than implies that the net core bootloader is included. Fix this.

File: cmake/mcuboot.cmake [hunk 4]
```diff
@@ -181,6 +181,34 @@ if(CONFIG_BOOTLOADER_MCUBOOT)
...
+    # bootloader is performed by MCUBoot on the application core. Hence we
+    # need a target for creating the signed binary of the network core
```
Comment (hakonfam) [inline cmake/mcuboot.cmake:187]: We don't need a target, we need to create the update.

File: cmake/mcuboot.cmake [hunk 5]
```diff
@@ -181,6 +181,34 @@ if(CONFIG_BOOTLOADER_MCUBOOT)
...
+    # application.
+
+    include(${CMAKE_BINARY_DIR}/hci_rpmsg/shared_vars.cmake)
```
Comment (hakonfam) [inline cmake/mcuboot.cmake:190]: Add a comment on why this is needed

File: cmake/mcuboot.cmake [hunk 6]
```diff
@@ -181,6 +181,34 @@ if(CONFIG_BOOTLOADER_MCUBOOT)
...
+
+    # TODO replace with proper domain once PR is in
```
Comment (hakonfam) [inline cmake/mcuboot.cmake]: https://github.com/nrfconnect/sdk-nrf/pull/2527

File: cmake/partition_manager.cmake [hunk 7]
```diff
@@ -358,6 +358,9 @@ if (is_dynamic_partition_in_domain)
   share("set(${domain}_PM_DOMAIN_HEADER_FILES ${header_files})")
   share("set(${domain}_PM_DOMAIN_IMAGES ${prefixed_images})")
   share("set(${domain}_PM_HEX_FILE ${PROJECT_BINARY_DIR}/${merged}.hex)")
+  share("set(${domain}_PM_DOTCONF_FILES ${pm_out_dotconf_files})")
+  # The partition manager script ensures that the 'app' hex always exists.
```
Comment (hakonfam) [inline cmake/partition_manager.cmake]: Remove this comment, it gives no info

File: cmake/partition_manager.cmake [hunk 8]
```diff
@@ -382,6 +385,51 @@ else()
...
+
+      foreach(name ${${d}_pm_var_names})
+        set_property(
+          TARGET partition_manager
+          PROPERTY ${d}_${name}
+          ${${name}}
+          )
+      endforeach()
+
+      # TODO don't check BOARD, don'c check for BT_RPMSG. Do something smarter, look at domains?
```
Comment (hakonfam) [inline cmake/partition_manager.cmake]: https://github.com/nrfconnect/sdk-nrf/pull/2527

Also, this is no longer a BOARD check, but SoC, which is OK

File: include/dfu/pcd.h [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+#define PCD_H__
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+#include <device.h>
+
+/** Magic value written to indicate that a copy should take place. */
+#define PCD_CMD_MAGIC_COPY 0xb5b4b3b6
+/** Magic value written to indicate that a something failed. */
```
Comment (hakonfam) [inline include/dfu/pcd.h:27]:
```suggestion
/** Magic value written to indicate that something failed. */
```

File: include/dfu/pcd.h [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+ *
+ *  The command is used to communicate information between the sender
+ *  and the receiver of the DFU image.
+ */
+struct pcd_cmd {
+	uint32_t magic;   /* Magic value to identify this structure in memory */
+	void * src_addr;  /* Source address to copy from */
+	size_t len;       /* Number of bytes to copy */
+	size_t offset;    /* Offset to store the flash image in */
+} __attribute__ ((aligned(4)));
```
Comment (hakonfam) [inline include/dfu/pcd.h:42]: Is this how alignment is done elsewhere in NCS?

File: include/dfu/pcd.h [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+#define APP_CORE_SRAM_BASE_ADDRESS 0x20000000
+#define APP_CORE_SRAM_SIZE KB(512)
+#if !defined(CONFIG_NRF_SPU_RAM_REGION_SIZE) //&& defined nrf53?
+#define CONFIG_NRF_SPU_RAM_REGION_SIZE 0x2000
+#endif
+#define PCD_CMD_ADDRESS (APP_CORE_SRAM_BASE_ADDRESS + APP_CORE_SRAM_SIZE \
```
Comment (hakonfam) [inline include/dfu/pcd.h:49]: Add some comments here, why does this need to be at this address? What is it used for?

File: include/dfu/pcd.h [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+struct pcd_cmd {
+	uint32_t magic;   /* Magic value to identify this structure in memory */
+	void * src_addr;  /* Source address to copy from */
+	size_t len;       /* Number of bytes to copy */
+	size_t offset;    /* Offset to store the flash image in */
+} __attribute__ ((aligned(4)));
+
+#define APP_CORE_SRAM_BASE_ADDRESS 0x20000000
+#define APP_CORE_SRAM_SIZE KB(512)
+#if !defined(CONFIG_NRF_SPU_RAM_REGION_SIZE) //&& defined nrf53?
```
Comment (hakonfam) [inline include/dfu/pcd.h:46]: When is this the case?

spm.c use 

```
/* SPU RAM regions */
#define RAM_SECURE_ATTRIBUTION_REGION_SIZE CONFIG_NRF_SPU_RAM_REGION_SIZE
#define NUM_RAM_SECURE_ATTRIBUTION_REGIONS (TOTAL_RAM_SIZE \
				/ RAM_SECURE_ATTRIBUTION_REGION_SIZE)
```

File: include/dfu/pcd.h [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+#define PCD_CMD_MAGIC_FAIL 0x25bafc15
+/** Magic value written to indicate that a copy is done. */
+#define PCD_CMD_MAGIC_DONE 0xf103ce5d
+
+/** @brief PCD command structure.
+ *
+ *  The command is used to communicate information between the sender
+ *  and the receiver of the DFU image.
+ */
+struct pcd_cmd {
```
Comment (hakonfam) [inline include/dfu/pcd.h:37]: Could we move this inside the  C file, and only have an empty struct definition here? To protect the members, and then update the signature of the `pcd_transfer` method to instead take src, len and offset as arguments.

File: samples/bluetooth/peripheral_uart/prj.conf [hunk 14]
```diff
@@ -51,3 +51,5 @@ CONFIG_LOG_BACKEND_UART=n
 
 CONFIG_ASSERT=y
 CONFIG_HW_STACK_PROTECTION=y
+
+CONFIG_BOOTLOADER_MCUBOOT=y
```
Comment (hakonfam) [inline samples/bluetooth/peripheral_uart/prj.conf:55]:
```suggestion
```

File: samples/nrf5340/netboot/README.rst [hunk 15]
```diff
@@ -0,0 +1,73 @@
+.. _nc_bootloader:
+
+nRF5340 Network core bootloader
+################################
+
+This bootloader sample implements an immutable first stage bootloader that has the capability to download update the application firmware on the network core of the nRF5340 SoC.
```
Comment (hakonfam) [inline samples/nrf5340/netboot/README.rst:6]:
```suggestion
This bootloader sample implements an immutable first stage bootloader that has the capability to update the application firmware on the network core of the nRF5340 SoC.
```

File: samples/nrf5340/netboot/README.rst [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+     Locking is done using the ACL peripheral.
+     For details on locking, see the :ref:`fprotect_readme` driver.
+
+
+#. Check if there is a pending firmware upgrade of the application.
+     Invoke the :ref:`subsys_pcd` subsys to inspect a memory region shared with the application core.
+     If the application core has configured a pending update, copy this to the application partition.
+     Once the copy is done, compare the SHA of the data in the application partition against the SHA provisioned in the firmware hex.
+
+#. Lock the flash of the application.
```
Comment (hakonfam) [inline samples/nrf5340/netboot/README.rst:28]:
```suggestion
#. Lock the flash of the application partition.
```

File: subsys/pcd/src/pcd.c [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+		return rc;
+	}
+
+	LOG_INF("Transfer done");
+	/* Signal complete by setting magic to DONE */
+	rsp->magic = PCD_CMD_MAGIC_DONE;
+	/* Wait for being rebooted */
+	while(1);
+	CODE_UNREACHABLE;
+	return 0;
```
Comment (hakonfam) [inline subsys/pcd/src/pcd.c]: This should be outside this function

**Review by hakonfam — CHANGES_REQUESTED** [review]

File: include/dfu/pcd.h [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+	uint32_t magic;   /* Magic value to identify this structure in memory */
+	void * src_addr;  /* Source address to copy from */
+	size_t len;       /* Number of bytes to copy */
+	size_t offset;    /* Offset to store the flash image in */
+} __attribute__ ((aligned(4)));
+
+#define APP_CORE_SRAM_BASE_ADDRESS 0x20000000
```
Comment (hakonfam) [inline include/dfu/pcd.h:44]: DT?

File: subsys/pcd/src/pcd.c (hunk 17 above)
Comment (sigvartmh) [reply subsys/pcd/src/pcd.c]: Fixed.

**Pull request closed without merging**